src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -618,25 +630,37 @@ private void ExecuteClusterTransactionOneByOne(List<ClusterTransactionCommand.Si
             try
             {
                 var index = command.Index;
+
                 var options = mergedCommands.Options[index];
                 if (exception == null)
                 {
-                    Task indexTask = null;
-                    if (options.WaitForIndexesTimeout != null)
-                    {
-                        indexTask = BatchHandler.WaitForIndexesAsync(DocumentsStorage.ContextPool, this, options.WaitForIndexesTimeout.Value,
-                            options.SpecifiedIndexesQueryString, options.WaitForIndexThrow,
-                            mergedCommands.LastDocumentEtag, mergedCommands.LastTombstoneEtag, mergedCommands.ModifiedCollections);
-                    }
 
-                    var result = new BatchHandler.ClusterTransactionCompletionResult
+                    var indexTask = BatchHandler.WaitForIndexesAsync(DocumentsStorage.ContextPool, this, options.WaitForIndexesTimeout,
+                        options.SpecifiedIndexesQueryString, options.WaitForIndexThrow,
+                        mergedCommands.LastDocumentEtag, mergedCommands.LastTombstoneEtag, mergedCommands.ModifiedCollections);
+
+                    var removeTask = ClusterTransactionWaiter.CreateTask(command.Options.TaskId, command.Index, out _);
+
+                    indexTask.ContinueWith(t =>
                     {
-                        Array = mergedCommands.Replies[index],
-                        IndexTask = indexTask,
-                    };
-                    ClusterTransactionWaiter.SetResult(options.TaskId, index, result);
+                        try
+                        {
+                            t.GetAwaiter().GetResult(); // Task t is already completed here
+                            ClusterTransactionWaiter.SetResult(options.TaskId, index, command.PreviousCount);
+                        }
+                        catch (Exception e)
+                        {
+                            ClusterTransactionWaiter.SetException(options.TaskId, index, e);
+                        }
+
+                        ThreadingHelper.InterlockedExchangeMax(ref LastCompletedClusterTransactionIndex, index);
```

Comment:
this sets that for all task, even faulty ones?

Reply:
Right, if the indexing failed it doesn't mean that the cluster transaction failed.